### PR TITLE
[primTorch] slice and transpose & etc.

### PR DESCRIPTION
**Pull request by mruberry**
This PR...

Adds the following prims:
- slice
- slice_in_dim
- transpose

Adds the following refs:
- cat
- permute
- transpose
- swap_axes (alias for transpose)
- tensor_split

Makes the following test improvements:
- adds reference inputs for torch.permute
- adds a NumPy reference for torch.permute
- adds reference inputs for torch.cat

Fixes the following bugs:
- adds support for scalars to the min and max prims

### CODE REVIEW

File: torch/_prims/utils.py [hunk 1]
```diff
@@ -549,11 +581,14 @@ def check_same_dtype(*args):
...
 
 
-def wrap_scalar(a: Number) -> torch.Tensor:
+def wrap_scalar(a: NumberType, *, dtype: torch.dtype = None) -> torch.Tensor:
     """
     Wraps a Number into a Tensor of corresponding dtype.
     """
-    return torch.tensor(a, dtype=type_to_dtype(type(a)))
+    if dtype is None:
+        return torch.tensor(a, dtype=type_to_dtype(type(a)))
```
Comment (ngimel) [inline torch/_prims/utils.py:589]: It's confusing that now type promotion is split between ref (that converts tensor) and here (converts scalar). Maybe wrapping indeed should be done in ref.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -932,3 +951,123 @@ def sum(
...
+
+
+def cat(
+    tensors: TensorSequenceType, dim: int = 0, out: TensorLikeType = None
+) -> TensorLikeType:
+    if len(tensors) == 0:
+        msg = "cat expects at least one tensor, but received zero!"
+        raise ValueError(msg)
+
+    _dim = utils.canonicalize_dims(tensors[0].ndim, dim)
```
Comment (ngimel) [inline torch/_refs/__init__.py:963]: why not canonicalize_idx?

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -932,3 +951,123 @@ def sum(
...
+    dtype, _ = _elementwise_dtypes(
+        *tensors, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+    )
+
+    _tensors = _convert_dtype(*tensors, dtype=dtype)
+    result = prims.concatenate(_tensors, _dim)
+
+    if out is not None:
+        out = _maybe_resize_out(out, result.shape)
+        return copy_to(out, result, allow_cross_device=False)  # type: ignore[arg-type]
```
Comment (ngimel) [inline torch/_refs/__init__.py:973]: Isn't it weird that all prims are not type-promoting except copy_to that is? Would it be better to instead check `out` dtype here, and not rely on `copy_to` to error out? If refs are documentation in code, then making error conditions clear is one of the goals.

File: torch/_prims/__init__.py [hunk 4]
```diff
@@ -636,16 +643,40 @@ def _lt_nvfuser(fd: Any, a: TensorLikeType, b: TensorLikeType):
     type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.ALWAYS_BOOL,
 )
 
+# Note: the following impls are because torch.maximum and torch.mininum do not support scalar inputs
+def _max_aten(
+    a: Union[TensorLikeType, NumberType], b: Union[TensorLikeType, NumberType]
+) -> TensorLikeType:
+    if isinstance(a, TensorLike) and isinstance(b, Number):
+        b = utils.wrap_scalar(b, dtype=a.dtype)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:651]: This happens after those checks, though

File: torch/_prims/__init__.py [hunk 5]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+) -> TensorLikeType:
+    _strides = strides if strides is not None else [1] * len(start_indices)
+
+    if a.ndim != len(start_indices):
+        msg = "Attempting to slice tensor of rank {0} with start_indices of length {1}!".format(
+            a.ndim, len(start_indices)
+        )
+        raise ValueError(msg)
+
+    if a.ndim != len(start_indices):
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:922]: Correct, this should be limit_indices here

I think the prim should do its own error checking and in the future we should create an error context so the name of the called function appears in the error message, and try to minimize redundant error checking in the ref

[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:922]: (from a trace perspective only prim error checking matters)

File: torch/_prims/__init__.py [hunk 6]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+        if x < 0:
+            msg = (
+                "Attempting to slice a tensor with a negative end index of {0}!".format(
+                    x
+                )
+            )
+            raise ValueError(msg)
+        if x > y:
+            msg = (
+                "Attempting to slice a tensor but a stop index in {0} is greater than the length of "
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:961]: In the Python slice docs it's start, stop and step: https://docs.python.org/3/c-api/slice.html

File: torch/_prims/__init__.py [hunk 7]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+                " its corresponding dimension in shape {1}".format(
+                    limit_indices, a.shape
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:963]: idk

File: torch/_prims/__init__.py [hunk 8]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+            msg = (
+                "Attempting to slice a tensor with a non-positive step of {0}!".format(
+                    x
+                )
+            )
+            raise ValueError(msg)
+
+    new_shape = []
+    for x, y in zip(start_indices, limit_indices):
+        new_shape.append(y - x)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:984]: yes! Good catch

File: torch/_prims/__init__.py [hunk 9]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+        msg = "slice_in_dim: received a negative axis {0}".format(axis)
+        raise ValueError(msg)
+    if axis >= a.ndim:
+        msg = "slice_in_dim: axis {0} is greater or equal to the rank {1} of the tensor".format(
+            axis, a.ndim
+        )
+        raise ValueError(msg)
+
+    if start_index < 0:
+        msg = "slice_in_dim: received a negative start_index {0}".format(start_index)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:1046]: Ikr, I kinda wanted to see what it was like

File: torch/_prims/__init__.py [hunk 10]
```diff
@@ -868,6 +901,214 @@ def _collapse_view_aten(a: Tensor, start: int, end: int) -> Tensor:
...
+            limit_index, start_index
+        )
+        raise ValueError(msg)
+
+    if stride < 0:
+        msg = "slice_in_dim: received a non-positive stride of {0}!".format(stride)
+        raise ValueError(msg)
+
+    start_indices = [0] * a.ndim
+    limit_indices = [0] * a.ndim
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_prims/__init__.py:1072]: Yes! -- that's weird I thought I fixed that

File: torch/_prims/utils.py (hunk 1 above)
Comment (mruberry) [reply torch/_prims/utils.py:589]: Yeah I'm going to look at scalar handling more generally and ensure we're preserving scalars properly

File: torch/_refs/__init__.py (hunk 2 above)
Comment (mruberry) [reply torch/_refs/__init__.py:963]: Yep, might as well

File: torch/_refs/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/__init__.py:973]: Let's reconsider this when we apply out handling as a wrapper?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**